Source/ASTextNode2.mm

```diff
@@ -986,29 +998,36 @@ - (NSArray *)pointSizeScaleFactors
 
 - (void)_ensureTruncationText
 {
+  ASLockScopeSelf();
   if (_textContainer.truncationToken == nil) {
-    ASLockScopeSelf();
     _textContainer.truncationToken = [self _locked_composedTruncationText];
   }
 }
 
 - (void)setTruncationAttributedText:(NSAttributedString *)truncationAttributedText
 {
-  if (ASLockedSelfCompareAssignCopy(_truncationAttributedText, truncationAttributedText)) {
+  ASLockScopeSelf();
+  if (!ASObjectIsEqual(_truncationAttributedText, truncationAttributedText)) {
+    _truncationAttributedText = [truncationAttributedText copy];
     [self _invalidateTruncationText];
```

Comment:
Why this change? Is it so that we now call `_invalidateTruncationText` with the lock held?

Reply:
Yes, otherwise there is no transactional integrity.

Reply:
OK in cases such as this (side effect keeps lock), we can use the non-locked compare-assign macros, like so:

```
ASLockScopeSelf();
if (ASCompareAssignCopy(_truncationAttributedText, truncationATtributedText)) {
  [self _invalidateTruncationText];
}
```

I set it up to work either way!

Reply:
Great! This is definitely an improvement.

I'm having trouble getting the minutes to circle back to this; will try, maybe this weekend.